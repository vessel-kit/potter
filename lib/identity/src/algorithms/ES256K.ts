import * as bytes from '@stablelib/bytes';
import * as elliptic from 'elliptic';
import * as sha256 from '@stablelib/sha256';
import { IPrivateKey, ISigner } from '../private-key.interface';
import { AlgorithmKind } from '../algorithm-kind';
import { IPublicKey, ISignatureVerification } from '../public-key.interface';
import BN from 'bn.js';

const secp256k1Context = new elliptic.ec('secp256k1');

export class PublicKey implements IPublicKey, ISignatureVerification {
  readonly alg = AlgorithmKind.ES256K;
  constructor(readonly material: Uint8Array) {}

  async verify(message: Uint8Array, signature: Uint8Array): Promise<boolean> {
    try {
      const keyPair = secp256k1Context.keyFromPublic(this.material);
      const r = new BN(signature.slice(0, 32));
      const s = new BN(signature.slice(32, 64));
      const digest = sha256.hash(message);
      return keyPair.verify(digest, { r, s });
    } catch {
      return false;
    }
  }
}

export class PrivateKey implements IPrivateKey, ISigner {
  readonly alg = AlgorithmKind.ES256K;

  #keyPair: elliptic.ec.KeyPair;
  #publicKey: Uint8Array;

  constructor(material: Uint8Array) {
    this.#keyPair = secp256k1Context.keyFromPrivate(material);
    this.#publicKey = new Uint8Array(this.#keyPair.getPublic().encodeCompressed());
  }

  async publicKey(): Promise<PublicKey> {
    return new PublicKey(this.#publicKey);
  }

  async sign(message: Uint8Array): Promise<Uint8Array> {
    const digest = sha256.hash(message);
    const signature = this.#keyPair.sign(digest, { canonical: true });
    const r = new Uint8Array(signature.r.toArray('be', 32));
    const s = new Uint8Array(signature.s.toArray('be', 32));
    return bytes.concat(r, s);
  }
}
