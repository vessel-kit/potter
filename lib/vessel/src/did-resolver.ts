import { DIDDocument, Resolver } from "did-resolver";
import { IResolver } from "@vessel-kit/identity";
import ThreeIdResolver from "@ceramicnetwork/3id-did-resolver";
import CeramicClient from "@ceramicnetwork/ceramic-http-client";

export class DIDResolver implements IResolver {
  #resolver: Resolver;

  constructor() {
    const ceramic = new CeramicClient("https://ceramic.3boxlabs.com");
    const threeIdResolver = ThreeIdResolver.getResolver(ceramic as any);
    this.#resolver = new Resolver({
      ...threeIdResolver
    });
  }

  resolve(didUrl: string): Promise<DIDDocument> {
    return this.#resolver.resolve(didUrl);
  }
}
