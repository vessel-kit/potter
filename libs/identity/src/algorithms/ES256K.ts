import * as bytes from '@stablelib/bytes';
import * as elliptic from 'elliptic';
import { IPrivateKey, ISigner } from '../private-key.interface';
import { AlgorithmKind } from '../algorithm-kind';
import { IPublicKey } from '../public-key.interface';
import BN from 'bn.js';

const secp256k1Context = new elliptic.ec('secp256k1');

export class PublicKey implements IPublicKey {
  readonly kind = AlgorithmKind.ES256K;
  constructor(readonly material: Uint8Array) {}
}

export class PrivateKey implements IPrivateKey, ISigner {
  readonly alg = AlgorithmKind.ES256K;

  #keyPair: elliptic.ec.KeyPair;
  #publicKey: Uint8Array;

  constructor(material: Uint8Array) {
    this.#keyPair = secp256k1Context.keyFromPrivate(material);
    this.#publicKey = new Uint8Array(this.#keyPair.getPublic().encodeCompressed());
  }

  async publicKey(): Promise<PublicKey> {
    return new PublicKey(this.#publicKey);
  }

  async sign(message: Uint8Array): Promise<Uint8Array> {
    const signature = this.#keyPair.sign(message, { canonical: true });
    const r = new Uint8Array(signature.r.toArray('be', 32));
    const s = new Uint8Array(signature.s.toArray('be', 32));
    return bytes.concat(r, s);
  }
}

export function verifySignature(publicKey: IPublicKey, message: Uint8Array, signature: Uint8Array): boolean {
  try {
    const keyPair = secp256k1Context.keyFromPublic(publicKey.material);
    const r = new BN(signature.slice(0, 32));
    const s = new BN(signature.slice(32, 64));
    return keyPair.verify(message, { r, s });
  } catch {
    return false;
  }
}
