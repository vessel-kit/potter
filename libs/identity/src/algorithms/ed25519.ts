import { IPrivateKey } from '../private-key.interface';
import { ISigning } from '../signing.interface';
import { AlgorithmKind } from '../algorithm-kind';
import { IPublicKey } from '../public-key.interface';
import * as ed25519 from '@stablelib/ed25519';

export class PublicKey implements IPublicKey {
  readonly kind = AlgorithmKind.ed25519;
  constructor(readonly material: Uint8Array) {}
}

export class PrivateKey implements IPrivateKey, ISigning {
  readonly kind = AlgorithmKind.ed25519;
  #keyPair: ed25519.KeyPair;
  #publicKey: Uint8Array;

  constructor(material: Uint8Array) {
    this.#keyPair = ed25519.generateKeyPairFromSeed(material as Buffer);
    this.#publicKey = this.#keyPair.publicKey;
  }

  async publicKey(): Promise<PublicKey> {
    return new PublicKey(this.#publicKey);
  }

  async sign(message: Uint8Array): Promise<Uint8Array> {
    return ed25519.sign(this.#keyPair.secretKey, message);
  }
}

export function verifySignature(publicKey: PublicKey, message: Uint8Array, signature: Uint8Array): boolean {
  try {
    return ed25519.verify(publicKey.material, message, signature);
  } catch {
    return false;
  }
}
