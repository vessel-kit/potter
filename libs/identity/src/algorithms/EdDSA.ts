import { IPrivateKey, ISigner } from '../private-key.interface';
import { AlgorithmKind } from '../algorithm-kind';
import { IPublicKey, ISignatureVerification } from '../public-key.interface';
import * as ed25519 from '@stablelib/ed25519';

export class PublicKey implements IPublicKey, ISignatureVerification {
  readonly alg = AlgorithmKind.EdDSA;
  constructor(readonly material: Uint8Array) {}

  async verify(message: Uint8Array, signature: Uint8Array): Promise<boolean> {
    try {
      return ed25519.verify(this.material, message, signature);
    } catch {
      return false;
    }
  }
}

export class PrivateKey implements IPrivateKey, ISigner {
  readonly alg = AlgorithmKind.EdDSA;
  #keyPair: ed25519.KeyPair;
  #publicKey: Uint8Array;

  constructor(material: Uint8Array) {
    this.#keyPair = ed25519.generateKeyPairFromSeed(material as Buffer);
    this.#publicKey = this.#keyPair.publicKey;
  }

  async publicKey(): Promise<PublicKey> {
    return new PublicKey(this.#publicKey);
  }

  async sign(message: Uint8Array): Promise<Uint8Array> {
    return ed25519.sign(this.#keyPair.secretKey, message);
  }
}
