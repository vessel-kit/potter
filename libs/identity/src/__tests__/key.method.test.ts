import { Resolver } from 'did-resolver';
import * as keyMethod from '../key.method';
import * as ES256K from '../algorithms/ES256K';
import * as Ed25519 from '../algorithms/EdDSA';
import * as _ from 'lodash';

describe('resolver', () => {
  const material = new Uint8Array(_.times(32, () => 1));

  const resolver = new Resolver({
    ...keyMethod.getResolver(),
  });

  test('resolve ES256K key', async () => {
    const publicKey = new ES256K.PublicKey(material);
    const did = keyMethod.identifier(publicKey).toString();
    const didDocument = await resolver.resolve(did);
    expect(didDocument).toMatchSnapshot();
  });

  test('resolve EdDSA key', async () => {
    const publicKey = new Ed25519.PublicKey(material);
    const did = keyMethod.identifier(publicKey).toString();
    const didDocument = await resolver.resolve(did);
    expect(didDocument).toMatchSnapshot();
  });

  test('not resolve garbage key did', async () => {
    await expect(resolver.resolve('did:key:garbage')).rejects.toThrow();
  });
});
