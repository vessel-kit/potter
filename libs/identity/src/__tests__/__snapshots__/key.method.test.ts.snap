// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolver resolve ES256K key 1`] = `
Object {
  "@context": "https://w3id.org/did/v1",
  "assertionMethod": Array [
    "did:key:z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N#z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N",
  ],
  "authentication": Array [
    "did:key:z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N#z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N",
  ],
  "capabilityDelegation": Array [
    "did:key:z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N#z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N",
  ],
  "capabilityInvocation": Array [
    "did:key:z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N#z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N",
  ],
  "id": "did:key:z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N",
  "publicKey": Array [
    Object {
      "controller": "did:key:z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N",
      "id": "did:key:z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N#z6DtMrg4Kv51UMAM8vJcCLcRywJfEB4dpHVxPCR6qm6hSV3N",
      "publicKeyHex": "0101010101010101010101010101010101010101010101010101010101010101",
      "type": "Secp256k1VerificationKey2018",
    },
  ],
}
`;

exports[`resolver resolve EdDSA key 1`] = `
Object {
  "@context": "https://w3id.org/did/v1",
  "assertionMethod": Array [
    "did:key:z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U#z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U",
  ],
  "authentication": Array [
    "did:key:z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U#z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U",
  ],
  "capabilityDelegation": Array [
    "did:key:z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U#z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U",
  ],
  "capabilityInvocation": Array [
    "did:key:z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U#z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U",
  ],
  "id": "did:key:z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U",
  "publicKey": Array [
    Object {
      "controller": "did:key:z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U",
      "id": "did:key:z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U#z6LSbk6TfcGsgm1yEUdGxwqscTzF6JkKNfrySPPLYqh8Ti6U",
      "publicKeyBase58": "4vJ9JU1bJJE96FWSJKvHsmmFADCg4gpZQff4P3bkLKi",
      "type": "Ed25519VerificationKey2018",
    },
  ],
}
`;
